### PR TITLE
[ADMIN] TGUI Check Antagonists panel

### PR DESCRIPTION
**Pull request by ynot01**
# Document the changes in your pull request

Tested thoroughly, has all of the features that old check antags panel had, with some extra (VV button)

### CODE REVIEW

**Review by Bibby0110 — CHANGES_REQUESTED** [review]
DNM until I review and help fix this

**Comment by ghost** [comment]
Looks cool

**Comment by JamieD1** [comment]
> DNM until I review and help fix this

Get on with it

File: code/modules/antagonists/nukeop/nukeop.dm [hunk 1]
```diff
@@ -396,6 +396,25 @@
 		return "Syndicates"
 
 /datum/team/nuclear/antag_listing_entry()
+	// Team constructor
+	// [0: "TeamName", 1:[AntagConstructor, AntagConstructor], 2:[Name of Tracked, [ [Name, TRACK_REF], [Name, TRACK_REF] ] ] ]
```
Comment (Bibby0110) [inline code/modules/antagonists/nukeop/nukeop.dm:400]: Construct these as associative lists

list(list("TeamName" = the_team_name, and so on))

File: code/modules/antagonists/revolution/revolution.dm [hunk 2]
```diff
@@ -403,18 +406,32 @@
 	return result.Join()
 
 /datum/team/revolution/antag_listing_entry()
+	// Team constructor
+	// [0: "TeamName", 1:[AntagConstructor, AntagConstructor], 2:[Name of Tracked, [TRACK_REF, TRACK_REF]]]
+	var/list/team_construct = list(antag_listing_name(), list(), list(FALSE,list()))
```
Comment (Bibby0110) [inline code/modules/antagonists/revolution/revolution.dm:411]: associative please

File: code/modules/admin/check_antagonists.dm [hunk 3]
```diff
@@ -1,173 +1,123 @@
...
-
-//Builds table row for the antag
-// Jim (Status) FLW PM TP
+/// Antag constructor
 /datum/antagonist/proc/antag_listing_entry()
-	var/list/parts = list()
-	if(show_name_in_check_antagonists)
-		parts += "[antag_listing_name()]([name])"
+	var/list/antag_construct
+	if(owner.current) // [0:"Traitor", 1:"Joe Schmoe", 2:"(DEAD)", 3:"[0x2100b3a5]", 4:MOB?, 5:CLIENT?, 6:Antag/Datum]
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:20]: associative list

File: code/modules/admin/check_antagonists.dm [hunk 4]
```diff
@@ -1,173 +1,123 @@
...
 	// Jim (Status) FLW PM TP
 	// Joe (Status) FLW PM TP
 	//Disk:
 	// Deep Space FLW
-	var/list/parts = list()
-	parts += "<b>[antag_listing_name()]</b><br>"
-	parts += "<table cellspacing=5>"
+	
+	// Team constructor
+	// [0: "TeamName", 1:[AntagConstructor, AntagConstructor], 2:[Name of Tracked, [[Name, TRACK_REF], [Name, TRACK_REF]]]]
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:51]: associative list

File: code/modules/admin/check_antagonists.dm [hunk 5]
```diff
@@ -1,173 +1,123 @@
...
-	var/list/priority_sections = list()
+/// Called by /client/proc/check_antagonists(), creates the tgui window
+/datum/admins/proc/check_antagonists()
+	if(!SSticker.HasRoundStarted())
+		alert("The game hasn't started yet!")
+		return
+	if(!check_rights(0))
+		return
+	
+	var/datum/tgui_check_antags/tgui = new(usr)//create the datum
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:77]: is there a specific reason we can't keep the datum around?

File: code/modules/admin/check_antagonists.dm [hunk 6]
```diff
@@ -1,173 +1,123 @@
...
-	if(CONFIG_GET(keyed_list/continuous)[SSticker.mode.config_tag])
-		dat += ", <a href='?_src_=holder;[HrefToken()];toggle_midround_antag=1'>[CONFIG_GET(keyed_list/midround_antag)[SSticker.mode.config_tag] ? "creating replacement antagonists" : "not creating new antagonists"]</a><BR>"
+/datum/tgui_check_antags/New(user) //user can either be a client or a mob
+	if (user) //Prevents runtimes on datums being made without clients
+		setup(user)
+
+/// Sets holder, which is supposed to be the interacting admin. Insecure, but harmless as it is checked with usr later.
+/datum/tgui_check_antags/proc/setup(user) //H can either be a client or a mob
+	if (istype(user,/client))
+		var/client/user_client = user
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:106]: there's no need to typecast

File: code/modules/admin/check_antagonists.dm [hunk 7]
```diff
@@ -1,173 +1,123 @@
...
-		dat += "Living crew limit: <a href='?_src_=holder;[HrefToken()];alter_midround_life_limit=1'>[CONFIG_GET(number/midround_antag_life_check) * 100]% of crew alive</a><BR>"
-		dat += "If limits past: <a href='?_src_=holder;[HrefToken()];toggle_noncontinuous_behavior=1'>[SSticker.mode.round_ends_with_antag_death ? "End The Round" : "Continue As Extended"]</a><BR>"
-	dat += "<a href='?_src_=holder;[HrefToken()];end_round=[REF(usr)]'>End Round Now</a><br>"
-	dat += "<a href='?_src_=holder;[HrefToken()];delay_round_end=1'>[SSticker.delay_end ? "End Round Normally" : "Delay Round End"]</a><br>"
-	dat += "<a href='?_src_=holder;[HrefToken()];ctf_toggle=1'>Enable/Disable CTF</a><br>"
-	dat += "<a href='?_src_=holder;[HrefToken()];rebootworld=1'>Reboot World</a><br>"
-	dat += "<a href='?_src_=holder;[HrefToken()];check_teams=1'>Check Teams</a>"
+		var/mob/user_mob = user
+		holder = user_mob.client //if its a mob, assign the mob's client to holder
+	spawn(5 SECONDS)
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:111]: avoid using SPAWN, use a timer or similar

File: code/modules/admin/check_antagonists.dm [hunk 8]
```diff
@@ -204,15 +155,361 @@
...
+	// Lists of antag constructors
+	data["antags"] = all_antagonists // Every solo antag
+	data["antag_types"] = antagonist_types // Every solo antag type
+	data["priority_sections"] = priority_sections // "Gamemode hero" team antags
+	data["sections"] = sections // Every other team antag
+	return data
+
+/// Custom check rights proc that uses check_rights_for()
+/datum/tgui_check_antags/proc/check_rights(rights_required, show_msg=1)
+	var/client/rights = holder
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:221]: stop using these unnecessary variables? holder is already a /client, just use that

File: code/modules/admin/check_antagonists.dm [hunk 9]
```diff
@@ -204,15 +155,361 @@
...
+	var/mob/mob_user = rights.mob
+	if(mob_user != usr) // If the holder has been edited, this will catch it before any badness can be done
+		return
+	if(rights != usr.client)
+		return
+	if(rights && mob_user)
+		if (check_rights_for(rights, rights_required))
+			return 1
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:229]: TRUE/FALSE in this proc

File: code/modules/admin/check_antagonists.dm [hunk 10]
```diff
@@ -204,15 +155,361 @@
...
+	return 0
+
+/datum/tgui_check_antags/ui_act(action, params)
+	if(..())
+		return
+	var/client/rights = holder
+	var/datum/admins/admindatum = rights.holder
+	var/mob/mob_user = rights.mob
+
+	if(mob_user != usr) // If the holder has been edited, this will catch it before any badness can be done
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:242]: tbh if you already do this i don't see why you even need the holder var on the datum? can't you just use usr all the way through?

File: code/modules/admin/check_antagonists.dm [hunk 11]
```diff
@@ -204,15 +155,361 @@
...
+				return
+			if(EMERGENCY_AT_LEAST_DOCKED)
+				return
+			if(SSshuttle.emergency.mode == SHUTTLE_CALL)
+				SSshuttle.emergency.cancel()
+				log_admin("[key_name(mob_user)] sent the Emergency Shuttle back.")
+				message_admins(span_adminnotice("[key_name_admin(mob_user)] sent the Emergency Shuttle back."))
+		
+		if("edit_shuttle_time")
+			if(!check_rights(R_SERVER))
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:276]: weird that this is R_SERVER when calling the shuttle is R_ADMIN, no?

File: code/modules/admin/check_antagonists.dm [hunk 12]
```diff
@@ -204,15 +155,361 @@
...
+				return
+			if(params["time"] && text2num(params["time"]))
+				var/timer = text2num(params["time"])
+				SSshuttle.emergency.setTimer(timer*10)
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:280]:
```suggestion
				SSshuttle.emergency.setTimer(timer SECONDS)
```

File: code/modules/admin/check_antagonists.dm [hunk 13]
```diff
@@ -204,15 +155,361 @@
...
+				log_admin("[key_name(mob_user)] edited the Emergency Shuttle's timeleft to [timer] seconds.")
+				minor_announce("The emergency shuttle will reach its destination in [round(SSshuttle.emergency.timeLeft(600))] minutes.")
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:282]:
```suggestion
				minor_announce("The emergency shuttle will reach its destination in [round(SSshuttle.emergency.timeLeft(1 MINUTES))] minutes.") //timeLeft variable is divisor, aka we show in minutes remaining
```

File: code/modules/admin/check_antagonists.dm [hunk 14]
```diff
@@ -204,15 +155,361 @@
...
+		return
+	if(rights && mob_user)
+		if (check_rights_for(rights, rights_required))
+			return 1
+		else
+			if(show_msg)
+				to_chat(mob_user, "<font color='red'>Error: You do not have sufficient rights to do that. You require one of the following flags:[rights2text(rights_required," ")].</font>", confidential=TRUE)
+	return 0
+
+/datum/tgui_check_antags/ui_act(action, params)
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:235]: block proccalls

File: code/modules/admin/check_antagonists.dm [hunk 15]
```diff
@@ -204,15 +155,361 @@
...
+			message_admins(span_adminnotice("[key_name_admin(mob_user)] edited the midround antagonist system to end the round upon failure."))
+
+		if("end_round")
+			if(!check_rights(R_ADMIN))
+				return
+			message_admins(span_adminnotice("[key_name_admin(mob_user)] is considering ending the round."))
+			if(alert(mob_user, "This will end the round, are you SURE you want to do this?", "Confirmation", "Yes", "No") == "Yes")
+				if(alert(mob_user, "Final Confirmation: End the round NOW?", "Confirmation", "Yes", "No") == "Yes")
+					message_admins(span_adminnotice("[key_name_admin(mob_user)] has ended the round."))
+					SSticker.force_ending = 1 //Yeah there we go APC destroyed mission accomplished
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:349]: TRUE

File: code/modules/admin/check_antagonists.dm [hunk 16]
```diff
@@ -204,15 +155,361 @@
...
+					message_admins(span_adminnotice("[key_name_admin(mob_user)] decided against ending the round."))
+			else
+				message_admins(span_adminnotice("[key_name_admin(mob_user)] decided against ending the round."))
+		
+		if("delay_round")
+			if(!check_rights(R_ADMIN))
+				return
+			if(!SSticker.delay_end)
+				SSticker.admin_delay_notice = input(usr, "Enter a reason for delaying the round end", "Round Delay Reason") as null|text
+				if(isnull(SSticker.admin_delay_notice))
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:361]: Ignore if you intend for ""/empty string to be valid
```suggestion
				if(!SSticker.admin_delay_notice)
```

File: code/modules/admin/check_antagonists.dm [hunk 17]
```diff
@@ -204,15 +155,361 @@
...
+					return
+			else
+				SSticker.admin_delay_notice = null
+			SSticker.delay_end = !SSticker.delay_end
+			var/reason = SSticker.delay_end ? "for reason: [SSticker.admin_delay_notice]" : "."//laziness
+			var/msg = "[SSticker.delay_end ? "delayed" : "undelayed"] the round end [reason]"
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:367]:
```suggestion
			var/msg = "[SSticker.delay_end ? "delayed" : "undelayed"] the round end[reason]"
```

File: code/modules/admin/check_antagonists.dm [hunk 18]
```diff
@@ -204,15 +155,361 @@
...
+			if(!check_rights(R_ADMIN))
+				return
+			if(!SSticker.delay_end)
+				SSticker.admin_delay_notice = input(usr, "Enter a reason for delaying the round end", "Round Delay Reason") as null|text
+				if(isnull(SSticker.admin_delay_notice))
+					return
+			else
+				SSticker.admin_delay_notice = null
+			SSticker.delay_end = !SSticker.delay_end
+			var/reason = SSticker.delay_end ? "for reason: [SSticker.admin_delay_notice]" : "."//laziness
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:366]:
```suggestion
			var/reason = SSticker.delay_end ? " for reason: [SSticker.admin_delay_notice]" : "."//laziness
```

File: code/modules/admin/check_antagonists.dm [hunk 19]
```diff
@@ -204,15 +155,361 @@
...
+		if("toggle_ctf")
+			if(!check_rights(R_ADMIN))
+				return
+			toggle_all_ctf(mob_user)
+
+		if("trigger_reboot")
+			if(!check_rights(R_ADMIN))
+				return
+			var/confirm = alert(mob_user, "Are you sure you want to reboot the server?", "Confirm Reboot", "Yes", "No")
+			if(confirm == "No")
```
Comment (Bibby0110) [inline code/modules/admin/check_antagonists.dm:382]: redundant

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+      height={710}
+      theme={'ntos'}>
+      {act('loadsuccess')}
+      <Window.Content scrollable>
+        <Grid height={15}>
+          <Grid.Column>
+            <Section title="Round Status" fill>
+              {replacementmode ? 'Former Game Mode: ' : 'Current Game Mode: '} <Box inline bold>{mode}</Box>
+              {replacementmode ? 'Replacement Game Mode: ' : ''} <Box inline bold>{replacementmode ? replacementmode : ''}</Box>
+              <br />
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:50]: can't you get rid of these brs and use some padding/margin

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+              <br />
+              Round Duration:
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:52]: Not a box?

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+              <br />
+              <Box inline bold>{time}</Box>
+              <br />
+              <br />
+              <Box bold>Emergency Shuttle</Box>
+              {shuttlecalled ? '' : <Input
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:58]: syntax probably not correct, but something like this. don't use a ternary for this
```suggestion
              {shuttlecalled && (<Input
```

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 23]
```diff
@@ -0,0 +1,331 @@
...
+                selfClear
+                placeholder={shuttletime}
+                onChange={(e, value) => act('edit_shuttle_time', {
+                  time: value,
+                })} />}
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:63]:
```suggestion
                })} />)}
```

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+                content={shuttlecalled ? 'Call Shuttle' : 'Send Back'}
+                onClick={() => act(shuttlecalled ? 'callshuttle' : 'recallshuttle')} />
+            </Section>
+          </Grid.Column>
+          <Grid.Column size={1.3}>
+            <Section title="Continuous Round Status" fill>
+              <Button
+                content={continue1 ? 'Continue if antagonists die' : 'End on antagonist death'}
+                onClick={() => act(continue1 ? 'end_antag_death' : 'cont_antag_death')} />
+              {continue1 ? <Button
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:75]: see above with regards to ternary

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 25]
```diff
@@ -0,0 +1,331 @@
...
+                content={continue2 ? 'creating replacement antagonists' : 'not creating new antagonists'}
+                onClick={() => act(continue2 ? 'dont_create_new' : 'create_new')} /> : ''}
+              <br />
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:78]: bad br

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 26]
```diff
@@ -0,0 +1,331 @@
...
+              <br />
+              {continue1 && continue2 ? 'Time limit: ' : ''}
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:80]: you get it by now

File: tgui/packages/tgui/interfaces/CheckAntags.js [hunk 27]
```diff
@@ -0,0 +1,331 @@
...
+                content="Legacy Panel"
+                onClick={() => act("forcelegacy")} />
+            </Section>
+          </Grid.Column>
+          <Grid.Column size={3}>
+            <Section title="Antagonists" fill>
+              {priority_sections.map((teamdata, index) =>
+                (
+                  <Section title={teamdata[0]} key={teamdata.index}>
+                    {teamdata[1].map((antagdata, idx) =>
```
Comment (Bibby0110) [inline tgui/packages/tgui/interfaces/CheckAntags.js:148]: when you've redone the DM side as associative lists these will function like objects

**Review by Bibby0110 — CHANGES_REQUESTED** [review]

**Comment by Bibby0110** [comment]
Stale, reopen when work on it

**Pull request closed without merging**